src/Microsoft.Azure.SignalR/ClientInvocation/RoutedClientResultsManager.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Threading;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal sealed class RoutedClientResultsManager: IRoutedClientResultsManager
+    {
+        private readonly ConcurrentDictionary<string,  RoutedInvocation> _routedInvocations = new();
+        private readonly ConcurrentDictionary<string, List<string>> _serviceMapping = new();
+
+        public void AddInvocation(string connectionId, string invocationId, string callerServerId, string instanceId, CancellationToken cancellationToken)
+        {
+            cancellationToken.Register(() => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
+
+            var result = _routedInvocations.TryAdd(invocationId, new RoutedInvocation(connectionId, callerServerId));
+            Debug.Assert(result);
+
+            AddServiceMappingMessage(instanceId, invocationId);
+        }
+
+        public bool TryCompleteResult(string connectionId, CompletionMessage message)
+        {
+            if (_routedInvocations.TryGetValue(message.InvocationId, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+                return _routedInvocations.TryRemove(message.InvocationId!, out _);
```

Comment:
! is a null-forgiving postfix, why add it here while not L30?

Reply:
removed. This fault is caused by copying from AspNetCore SignalR